### ISSUE TITLE
`sqflite` adapter for Serverpod protocol classes

### ISSUE THREAD

**Issue by lukehutch** [serverpod/serverpod#1719]
I need to persist some data locally on my Flutter client, to reduce redundant requests for the same data to the server, and I want to be able to use the same interface for storing data in SQLite on the client (via [`sqflite`](https://pub.dev/packages/sqflite)) that I use for writing data to Postgres on the server.

It would be super awesome if Serverpod supported this out of the box (I'm sure I'll have to write my own version of this before you get to it, but it's an important feature to support generically).

What I plan to do is cache records on the client, and only download new/changed records as necessary from the server, then run all queries against the client copy of the data.

I want to be able to receive a list of protocol objects from the server, and write them directly to SQLite, with the data split out into columns, just as it is on the server.

It would be great if `serverpod generate` could generate a client (SQLite) variant of the Postgres SQL table creation code for any protocol objects (and fields thereof) that are sent over the wire (as opposed to being server-database-only), with appropriate indexes.

**Comment by Isakdl** [comment]
This is a pretty cool idea! 

We could build some smart persistence and offline support on top of this too. Such as the ability to save things offline and then sync when the device gets online again.

**Comment by lukehutch** [comment]
Yes, exactly. I was going to write a "poor man's sync" for append-only tables (eg chat history), where the difference in rows between the client and a user-specific view on the server is all that is needed to know how many additional rows to fetch, but probably a slightly more powerful sync capability would be possible without the complexity of full replication sync support.

**Comment by dkbast** [comment]
Since discussions don't seem to create backlinks I'll add it manually: https://github.com/serverpod/serverpod/discussions/2084 - the CRDT pattern could solve this as well as https://github.com/serverpod/serverpod/issues/2080 which is pretty much the same as this ticket but with a different database backend, as there is a CRDT implementation for both this could be a good way forward.

**Comment by marcelomendoncasoares** [comment]
We now have SQLite generated on the client since 3.5.0-beta, which will be out on the official 4.0.0.
